Mage.Sets/src/mage/cards/c/CaseOfTheBurningMasks.java

```diff
@@ -0,0 +1,206 @@
+
+    public CaseOfTheBurningMasks(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.ENCHANTMENT}, "{1}{R}{R}");
+
+        this.subtype.add(SubType.CASE);
+
+        // When this Case enters the battlefield, it deals 3 damage to target creature an opponent controls.
+        Ability initialAbility = new EntersBattlefieldTriggeredAbility(
+                new DamageTargetEffect(3, "it"));
+        initialAbility.addTarget(new TargetOpponentsCreaturePermanent());
+        initialAbility.addTarget(new TargetOpponentsCreaturePermanent());
+        // To solve -- Three or more sources you controlled dealt damage this turn.
+        Condition toSolveCondition = new CaseOfTheBurningMasksCondition();
+        // Solved -- Sacrifice this Case: Exile the top three cards of your library. Choose one of them. You may play that card this turn.
+        Ability solvedAbility = new ConditionalActivatedAbility(new CaseOfTheBurningMasksEffect(),
+                new SacrificeSourceCost(), SolvedSourceCondition.SOLVED, false);
+
+        this.addAbility(new CaseAbility(initialAbility, toSolveCondition, solvedAbility)
+                .addHint(CaseOfTheBurningMasksHint.instance),
+                new CaseOfTheBurningMasksWatcher());
+    }
+
+    private CaseOfTheBurningMasks(final CaseOfTheBurningMasks card) {
+        super(card);
+    }
+
+    @Override
+    public CaseOfTheBurningMasks copy() {
+        return new CaseOfTheBurningMasks(this);
+    }
+}
+
+class CaseOfTheBurningMasksCondition implements Condition {
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        CaseOfTheBurningMasksWatcher watcher = game.getState().getWatcher(CaseOfTheBurningMasksWatcher.class);
+        return watcher != null && watcher.damagingCountByController(source.getControllerId()) >= 3;
+    }
+
+    @Override
+    public String toString() {
+        return "Three or more sources you controlled dealt damage this turn";
+    }
+}
+
+enum CaseOfTheBurningMasksHint implements Hint {
+    instance;
+
+    @Override
+    public CaseOfTheBurningMasksHint copy() {
+        return this;
+    }
+
+    @Override
+    public String getText(Game game, Ability ability) {
+        Permanent permanent = game.getPermanent(ability.getSourceId());
+        if (permanent == null) {
+            return "";
+        }
+        if (permanent.isSolved()) {
+            return "Case is solved";
+        }
+        int sources = game.getState()
+                .getWatcher(CaseOfTheBurningMasksWatcher.class)
+                .damagingCountByController(ability.getControllerId());
+        StringBuilder sb = new StringBuilder("Case is unsolved. Sources that dealt damage this turn: ");
+        sb.append(sources);
+        sb.append(" (need 3).");
+        if (sources > 2 && game.isActivePlayer(ability.getControllerId())) {
+            sb.append(" Case will be solved at the end step.");
+        }
+        return sb.toString();
```

Comment:
this is a nice detailed hint. Do you think it's possible to generalize instead of implementing separately on each case, or do you think that would be too convoluted?

also fyi there's no point in using StringBuilder except in loops, concatenating strings is fine

Reply:
The boilerplate text (e.g. "Case is solved/unsolved") can be generalized, but I couldn't see a neat way of delegating the core value calculation and description text. If you have any suggestions, I'd be more than happy to take them on board.

Reply:
1. Can you make it extend `ConditionHint` so you can see at a glance if solved or unsolved with the icon?
2. Maybe can put common logic into an abstract class `CaseSolvedHint` that has one method to override for the condition-related text (params Game, Ability, Permanent).